plugin/gen_embedding/plugin.cc

```diff
@@ -0,0 +1,455 @@
+      api_key = THDVAR(thd, api_key);
+      uint max_dimensions = 3072; // Default to the largest embedding size
+      if (args[1]->const_item()) { // If a const, we can parse the model name here and determine the max dimensions 
+        std::string model_name= (std::string) args[1]->val_str()->c_ptr();
+        if (model_dimensions.find(args[1]->val_str()->c_ptr()) != model_dimensions.end()) {
+          max_dimensions = model_dimensions[model_name];
+        }
+        else {
+          // TODO: Can we also early exit here?
+          my_printf_error(1, "GENERATE_EMBEDDING_OPENAI: "
+            "Model %s is not supported", ME_ERROR_LOG | ME_WARNING, model_name.c_str());
+        }
+      }
+
+      decimals= 0;
+      MEM_ROOT *root= thd->active_stmt_arena_to_use()->mem_root;
+      uint n_paths = 1;
+      paths= (json_path_with_flags *) alloc_root(root, sizeof(json_path_with_flags) * n_paths);
+      // We want to store max_dimensions floats, each 4 bytes 
+      fix_length_and_charset(max_dimensions * 4, &my_charset_bin);
+      set_maybe_null();
+      return false;
+    }
+    
+    String *val_str(String *buf) override
+    {
+      json_value_types type;
+      char *value;
+
+      if (make_openai_request())
+      {
+        null_value= true;
+        return NULL;
+      }   
+      return read_json(buf, &type, &value);
+    }
+    
+    LEX_CSTRING func_name_cstring() const override
+    {
+      static LEX_CSTRING name= "GENERATE_EMBEDDING_OPENAI"_LEX_CSTRING;
+      return name;
+    }
+    Item *do_get_copy(THD *thd) const override
+    { return get_item_copy<Item_func_gen_embedding>(thd, this); }
+  
+  class Create_func : public Create_func_arg2
+  {
+  public:
+    Item *create_2_arg(THD *thd, Item *arg1, Item* arg2) override
+    {
+      return new (thd->mem_root) Item_func_gen_embedding(thd, arg1, arg2);
+    }
+  };
+
+  static Plugin_function *plugin_descriptor()
+  {
+    static Create_func creator;
+    static Plugin_function descriptor(&creator);
+    return &descriptor;
+  }
+};
+
+
+String *Item_func_gen_embedding::read_json(String *str,
+                                          json_value_types *type,
+                                          char **out_val)
+{
+  String *js= &api_response;
+  json_engine_t je;
+  json_path_t p;
+  const uchar *value;
+  int not_first_value= 0, count_path= 0;
+  size_t v_len;
+  int array_size_counter[JSON_DEPTH_LIMIT];
+  json_path_with_flags *c_path= paths;
+
+  if ((null_value= args[0]->null_value))
+    return 0;
+
+  c_path->p.types_used= JSON_PATH_KEY_NULL;
+  const uchar* s_p = (const uchar*) JSON_EMBEDDING_PATH.c_str();
+  if (s_p)
+  {
+    if (json_path_setup(&c_path->p,&my_charset_utf8mb4_general_ci, s_p,
+                      s_p + strlen((const char *) s_p)))
+    {
+    //  report_path_error(s_p, &c_path->p, n_arg);
+      goto return_null;
+    }
+  }
+
+  *type= JSON_VALUE_NULL;
+
+  if (str)
+  {
+    str->set_charset(js->charset());
```

Comment:
not really? `str` is the resulting vector, not json. it should be always `my_charset_bin`, as far as I understand.